dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedScoreTracker2.java

```diff
@@ -0,0 +1,343 @@
+         */
+        private final CoarseExponentialDecayReservoir recentFailuresReservoir;
+
+        ChannelScoreInfo(
+                int hostIndex,
+                ConcurrencyLimitedChannel concurrencyLimitedChannel,
+                Ticker clock,
+                PerHostObservability observability) {
+            this.hostIndex = hostIndex;
+            this.concurrencyLimitedChannel = concurrencyLimitedChannel;
+            this.recentFailuresReservoir = new CoarseExponentialDecayReservoir(clock::read, FAILURE_MEMORY);
+            this.observability = observability;
+        }
+
+        @Override
+        public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+            Optional<ListenableFuture<Response>> maybe = concurrencyLimitedChannel.maybeExecute(endpoint, request);
+            if (maybe.isPresent()) {
+                observability.markRequestMade();
+                DialogueFutures.addDirectCallback(maybe.get(), this);
+            }
+
+            return maybe;
+        }
+
+        @Override
+        public void onSuccess(Response response) {
+            if (Responses.isQosStatus(response) || Responses.isServerError(response)) {
+                recentFailuresReservoir.update(FAILURE_WEIGHT);
+                observability.debugLogStatusFailure(response);
+            } else if (Responses.isClientError(response)) {
+                // We track 4xx responses because bugs in the server might cause one node to erroneously
+                // throw 401/403s when another node could actually return 200s. Empirically, healthy servers
+                // do actually return a continuous background rate of 4xx responses, so we weight these
+                // drastically less than 5xx responses.
+                recentFailuresReservoir.update(FAILURE_WEIGHT / 100);
+                observability.debugLogStatusFailure(response);
+            }
+        }
+
+        public int channelIndex() {
+            return hostIndex;
+        }
+
+        @Override
+        public void onFailure(Throwable throwable) {
+            recentFailuresReservoir.update(FAILURE_WEIGHT);
+            observability.debugLogThrowableFailure(recentFailuresReservoir, throwable);
+        }
+
+        /** A low score is considered desirable (and they can be negative). */
+        private ScoreSnapshot computeScoreSnapshot() {
+            int requestsInflight = concurrencyLimitedChannel.getInflight();
+            int limit = (int) concurrencyLimitedChannel.getMax().orElse(0);
+
+            double failureReservoir = recentFailuresReservoir.get();
+
+            // it's important that scores are integers because if we kept the full double precision, then a single 4xx
+            // would end up influencing host selection long beyond its intended lifespan in the absence of other data.
+            int score = requestsInflight - limit + Ints.saturatedCast(Math.round(failureReservoir));
```

Comment:
Here's the interesting change:
```
-int score = requestsInflight + Ints.saturatedCast(Math.round(failureReservoir));
+int score = requestsInflight - limit + Ints.saturatedCast(Math.round(failureReservoir));
```

I just decided to linearly factor the concurrencylimiter's upper limit in... it's a bit confusing because low score = good, so many inflight = bad, but we want a channel with high available capacity to have a low score, so I just subtracted it :shrug:

Reply:
Assuming some base rate of failures (qos, network blips, etc) the in-flight requests and limit should be roughly equivalent under load.